File: scripts/prune_players.js
import { readFileSync, writeFileSync } from 'fs';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const src = join(__dirname, '..', 'public', 'players.json');
const out = join(__dirname, '..', 'public', 'players_pruned.json');

// Fields to keep (adjust if you need others)
const KEEP = new Set(['player_id', 'first_name', 'last_name', 'full_name', 'position', 'team', 'status', 'bye_week', 'fantasy_positions']);

function pruneObject(obj) {
  const res = {};
  for (const k of Object.keys(obj)) {
    if (KEEP.has(k)) res[k] = obj[k];
  }
  // ensure player_id exists if original key is id-like
  if (!res.player_id && (obj.id || obj.playerId)) {
    res.player_id = obj.id || obj.playerId;
  }
  return res;
}

try {
  const raw = readFileSync(src, 'utf8');
  const data = JSON.parse(raw);

  let pruned;
  if (Array.isArray(data)) {
    pruned = data.map(pruneObject);
  } else if (typeof data === 'object' && data !== null) {
    // handle map of id -> player
    const isMap = Object.values(data).every((v) => typeof v === 'object');
    if (isMap) {
      pruned = {};
      for (const [k, v] of Object.entries(data)) {
        pruned[k] = pruneObject(v);
        // keep key if player_id missing
        if (!pruned[k].player_id) pruned[k].player_id = k;
      }
    } else {
      pruned = pruneObject(data);
    }
  } else {
    throw new Error('Unexpected players.json format');
  }

  writeFileSync(out, JSON.stringify(pruned, null, 2));
  console.log('Pruned players written to', out);
} catch (err) {
  console.error('Error:', err.message);
  process.exit(1);
}

File: scripts/prune_players.py
import json
import sys
from pathlib import Path
from urllib.request import urlopen, Request
from urllib.error import URLError

# Setup paths relative to this script
# Equivalent to: join(__dirname, '..', 'public', '...')
BASE_DIR = Path(__file__).resolve().parent.parent / "public"
OUT_FILE = BASE_DIR / "players_pruned.json"

# Ensure directory exists
BASE_DIR.mkdir(parents=True, exist_ok=True)

# Fields to keep
KEEP = {
    "player_id",
    "first_name",
    "last_name",
    "full_name",
    "position",
    "team",
    "status",
    "bye_week",
    "fantasy_positions",
    "injury_status",
}

# Allowed positions
ALLOWED_POSITIONS = {"QB", "RB", "TE", "WR"}


def prune_object(obj):
    """Creates a new dictionary containing only the allowed keys."""
    res = {k: v for k, v in obj.items() if k in KEEP}

    # Ensure player_id exists if original key is id-like
    if "player_id" not in res:
        if "id" in obj:
            res["player_id"] = obj["id"]
        elif "playerId" in obj:
            res["player_id"] = obj["playerId"]

    return res


def should_keep_player(player):
    """Checks if player has an allowed position and is not Inactive or without a team."""
    position = player.get("position")
    status = player.get("status")
    team = player.get("team")
    return position in ALLOWED_POSITIONS and status != "Inactive" and team is not None


def main():
    print("Downloading players data...")

    try:
        # Fetch data using standard library
        with urlopen(Request("https://api.sleeper.app/v1/players/nfl")) as response:
            data = json.loads(response.read().decode())

        # Prune Data (keep in-memory, no raw file on disk)
        pruned_data = None

        if isinstance(data, list):
            pruned_data = [
                prune_object(item) for item in data if should_keep_player(item)
            ]

        elif isinstance(data, dict):
            # Handle map of id -> player (Sleeper API usually returns this)
            # Check if values are dicts to confirm it's a map
            if data and isinstance(next(iter(data.values())), dict):
                pruned_data = {}
                for k, v in data.items():
                    if should_keep_player(v):
                        pruned_item = prune_object(v)
                        # Keep key as player_id if missing
                        if "player_id" not in pruned_item:
                            pruned_item["player_id"] = k
                        pruned_data[k] = pruned_item
            else:
                # Single object
                pruned_data = prune_object(data)
        else:
            raise ValueError("Unexpected players.json format")

        # 3. Write Pruned Data
        with open(OUT_FILE, "w", encoding="utf-8") as f:
            json.dump(pruned_data, f, indent=2, sort_keys=True)
        print(f"Pruned players written to {OUT_FILE}")

    except (URLError, ValueError, OSError) as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()
